### PR TITLE
Spring security

### PR DESCRIPTION
**Pull request by Cisfinitum**
So, here we have bunch: spring security + h2 database
First of all u ll see main page, where u can log into our system. If there are no such name and password in our databse u ll be redirected to the page 403 with message "feels bad, man". Otherwise u ll be redirected to the page "testpage". There are two links on dat page. "admin" and "user". According to our access status, u ll have different messages on both links.
Also, the scheme of that project is complicated, so i ll explain how dat code works on nearest meeting.

### CODE REVIEW

File: src/main/java/com/epam/service/PersonDetailsServiceImpl.java [hunk 1]
```diff
@@ -18,17 +19,21 @@
...
     @Override
-    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
-        User user = userService.getUser("user");
-        UserRoleEnum currole = user.getRole();
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        Person person = personService.getPerson(username);
+        PersonRoleEnum personRoleEnum = person.getRole();
         Set<GrantedAuthority> roles = Stream.of(new SimpleGrantedAuthority
-                (Objects.requireNonNullElse(currole, UserRoleEnum.ANONYMOUS).getEnumrole()))
+                (Objects.requireNonNullElse(personRoleEnum, PersonRoleEnum.ANONYMOUS).getEnumrole()))
```
Comment (goponenkoev) [inline src/main/java/com/epam/service/PersonDetailsServiceImpl.java]: We use jdk 8, please fix this line

File: src/main/java/com/epam/service/PersonService.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package com.epam.service;
+
```
Comment (goponenkoev) [inline src/main/java/com/epam/service/PersonService.java]: Unnecessary empty lines

File: src/main/java/com/epam/controller/MainController.java [hunk 3]
```diff
@@ -1,54 +1,47 @@
...
 @Controller
+@RequestMapping("/")
 public class MainController {
-  @GetMapping("/index")
-  public String defaultPage(Principal user, ModelMap model) {
-    if (user != null) {
-      model.addAttribute("name", user.getName());
-    } else {
-      model.addAttribute("name", "guest");
+
```
Comment (goponenkoev) [inline src/main/java/com/epam/controller/MainController.java]: And here

File: src/main/java/com/epam/model/PersonRoleEnum.java [hunk 4]
```diff
@@ -1,14 +1,16 @@
-package com.epam.entity.enums;
+package com.epam.model;
+
```
Comment (goponenkoev) [inline src/main/java/com/epam/model/PersonRoleEnum.java]: And here

File: src/main/java/com/epam/repository/PersonDAO.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
+package com.epam.repository;
+
+import com.epam.model.Person;
+import com.epam.model.PersonRoleEnum;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+
```
Comment (goponenkoev) [inline src/main/java/com/epam/repository/PersonDAO.java]: And here

File: src/main/java/com/epam/service/PersonDetailsServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import com.epam.model.PersonRoleEnum;
+import lombok.Setter;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
```
Comment (goponenkoev) [inline src/main/java/com/epam/service/PersonDetailsServiceImpl.java]: And here

File: src/main/java/com/epam/service/PersonService.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+
+import com.epam.model.Person;
+import com.epam.repository.PersonDAO;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
```
Comment (goponenkoev) [inline src/main/java/com/epam/service/PersonService.java]: And here

File: src/main/java/com/epam/model/Person.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+@AllArgsConstructor
+@Builder
+@Setter
+@Getter
+public class Person {
+    private Integer id;
+    private String name;
+    private String password;
+    private PersonRoleEnum role;
+
```
Comment (goponenkoev) [inline src/main/java/com/epam/model/Person.java]: And here...

File: src/main/java/com/epam/repository/PersonDAO.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    private final NamedParameterJdbcTemplate jdbcTemplate;
+
+    @Autowired
+    public PersonDAO(NamedParameterJdbcTemplate jdbcTemplate) {
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    public List<Person> getPersons(){
+        return jdbcTemplate.query("SELECT * FROM person",
+                (ResultSet rs, int rowNum) -> Person.builder()
```
Comment (imbirevkotlya) [inline src/main/java/com/epam/repository/PersonDAO.java]: please, remove ResultSet and int, and use lambda type method reference here (see discussion of prev request)

File: src/main/java/com/epam/service/PersonService.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+public class PersonService {
+
+    private final PersonDAO personDAO;
+    @Autowired
+    public PersonService(PersonDAO personDAO) {
+        this.personDAO = personDAO;
+    }
+
+    public void printPersonsNames(){
+        List<Person> persons = personDAO.getPersons();
```
Comment (imbirevkotlya) [inline src/main/java/com/epam/service/PersonService.java]: Why persons and not people?

File: build.gradle [hunk 11]
```diff
@@ -27,14 +27,16 @@ dependencies {
     compile group: 'org.springframework.security', name: 'spring-security-web', version: '5.1.3.RELEASE'
     compileOnly "org.projectlombok:lombok:1.18.4"
     testCompileOnly "org.projectlombok:lombok:1.18.4"
-    compile group: 'org.springframework', name: 'spring-webmvc', version: '5.1.4.RELEASE'
+    compile group: 'org.springframework', name: 'spring-webmvc', version: '5.1.3.RELEASE'
```
Comment (imbirevkotlya) [inline build.gradle]: please, use variable for spring version

File: src/main/java/com/epam/model/Person.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
+package com.epam.model;
+
+import lombok.*;
+
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+@Setter
+@Getter
```
Comment (imbirevkotlya) [inline src/main/java/com/epam/model/Person.java]: Only @Data?

File: src/main/java/com/epam/service/PersonDetailsServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Setter
```
Comment (imbirevkotlya) [inline src/main/java/com/epam/service/PersonDetailsServiceImpl.java]: Do we need setter here?

File: src/main/java/com/epam/service/PersonService.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+        persons.forEach(person -> System.out.println(person.getName()));
+    }
+
+    public Person getPerson(String name){
```
Comment (imbirevkotlya) [inline src/main/java/com/epam/service/PersonService.java]: Read about DRY principle: you have already had good method above, try to use it here

Comment (imbirevkotlya) [reply src/main/java/com/epam/service/PersonService.java]: What will be if we pass null as an argument?

File: src/main/java/com/epam/service/PersonDetailsServiceImpl.java (hunk 1 above)
Comment (Cisfinitum) [reply src/main/java/com/epam/service/PersonDetailsServiceImpl.java]: Done

File: src/main/java/com/epam/service/PersonService.java (hunk 2 above)
Comment (Cisfinitum) [reply src/main/java/com/epam/service/PersonService.java]: Done

File: src/main/java/com/epam/controller/MainController.java (hunk 3 above)
Comment (Cisfinitum) [reply src/main/java/com/epam/controller/MainController.java]: Done

File: src/main/java/com/epam/model/PersonRoleEnum.java (hunk 4 above)
Comment (Cisfinitum) [reply src/main/java/com/epam/model/PersonRoleEnum.java]: Done

File: src/main/java/com/epam/repository/PersonDAO.java (hunk 5 above)
Comment (Cisfinitum) [reply src/main/java/com/epam/repository/PersonDAO.java]: Done

File: src/main/java/com/epam/service/PersonDetailsServiceImpl.java (hunk 6 above)
Comment (Cisfinitum) [reply src/main/java/com/epam/service/PersonDetailsServiceImpl.java]: Done

File: src/main/java/com/epam/model/Person.java (hunk 8 above)
Comment (Cisfinitum) [reply src/main/java/com/epam/model/Person.java]: Done

File: src/main/java/com/epam/repository/PersonDAO.java (hunk 9 above)
Comment (Cisfinitum) [reply src/main/java/com/epam/repository/PersonDAO.java]: Done

File: src/main/java/com/epam/service/PersonService.java (hunk 10 above)
Comment (Cisfinitum) [reply src/main/java/com/epam/service/PersonService.java]: Done

File: build.gradle (hunk 11 above)
Comment (Cisfinitum) [reply build.gradle]: Done

File: src/main/java/com/epam/service/PersonDetailsServiceImpl.java (hunk 13 above)
Comment (Cisfinitum) [reply src/main/java/com/epam/service/PersonDetailsServiceImpl.java]: Done

**Comment by imbirevkotlya** [comment]
commit messages are perfect))))

File: src/test/java/com/epam/service/PersonDetailsServiceImplTest.java [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.mockito.Mockito.when;
+import static org.junit.jupiter.api.Assertions.*;
+
+class PersonDetailsServiceImplTest {
```
Comment (goponenkoev) [inline src/test/java/com/epam/service/PersonDetailsServiceImplTest.java]: Hmm, I pretty sure this class should be named smth like this: `PersonServiceTest`, what do u think?

File: src/test/java/com/epam/service/PersonDetailsServiceImplTest.java [hunk 16]
```diff
@@ -0,0 +1,55 @@
+package com.epam.service;
+
+import com.epam.model.Person;
+import com.epam.model.PersonRoleEnum;
+import com.epam.repository.PersonDAO;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
```
Comment (goponenkoev) [inline src/test/java/com/epam/service/PersonDetailsServiceImplTest.java]: And also please delete here blank line

File: src/test/java/com/epam/service/PersonDetailsServiceImplTest.java (hunk 15 above)
Comment (Cisfinitum) [reply src/test/java/com/epam/service/PersonDetailsServiceImplTest.java]: Yeup, u re right. Fixed

File: src/test/java/com/epam/service/PersonDetailsServiceImplTest.java (hunk 16 above)
Comment (Cisfinitum) [reply src/test/java/com/epam/service/PersonDetailsServiceImplTest.java]: Done

**Pull request merged**